src/v/cluster/cluster_epoch_generator.cc

```diff
@@ -0,0 +1,130 @@
+
+namespace cluster {
+
+namespace {
+
+using namespace std::chrono_literals;
+
+// TODO(cloud-topics): make these configuration knobs
+constexpr ss::lowres_clock::duration epoch_cache_timeout = 1s;
+// The maximum amount of time to tolerate the same epoch being returned
+constexpr ss::lowres_clock::duration max_same_epoch_duration = 1min;
+
+// NOLINTNEXTLINE(*-avoid-reference-coroutine-parameters)
+ss::future<> force_increment_loop(controller_stm& stm, ss::abort_source& as) {
+    model::offset last_offset = co_await stm.bootstrap_committed_offset();
+    while (!as.abort_requested()) {
+        try {
+            co_await ss::sleep_abortable<ss::lowres_clock>(
+              max_same_epoch_duration, as);
+        } catch (const ss::sleep_aborted&) {
+            co_return;
+        }
+        model::offset new_offset = co_await stm.bootstrap_committed_offset();
+        if (new_offset > last_offset) {
+            last_offset = new_offset;
+            continue;
+        }
+        auto result = co_await stm.quorum_write_empty_batch(
+          ss::lowres_clock::now() + 3s);
+        if (result) {
+            last_offset = result.value().last_offset;
+            vlog(
+              clusterlog.debug,
+              "Forced increment of cluster epoch to: {}",
+              last_offset);
+        } else if (result.error() != raft::errc::not_leader) {
+            vlog(
+              clusterlog.warn,
+              "Failed to force increment epoch: {}",
+              result.error());
+        }
+    }
+}
+
+} // namespace
+
+struct cluster_epoch_generator::shard0_state {
+    ss::sharded<controller_stm>& stm;
+    ss::future<> force_increment_loop;
+    ss::abort_source abort_source;
+};
+
+cluster_epoch_generator::cluster_epoch_generator() noexcept = default;
+cluster_epoch_generator::~cluster_epoch_generator() noexcept = default;
+
+ss::future<> cluster_epoch_generator::start() { co_return; }
+ss::future<> cluster_epoch_generator::stop() {
+    co_await _gate.close();
+    if (_shard0_state) {
+        _shard0_state->abort_source.request_abort();
+        co_await std::move(_shard0_state->force_increment_loop);
+        _shard0_state = nullptr;
+    }
+}
+
+ss::future<int64_t> cluster_epoch_generator::current_epoch() {
+    auto holder = _gate.hold();
+    if (cache_entry_expired()) {
+        auto units = co_await _mu.get_units();
+        if (cache_entry_expired()) {
+            int64_t new_epoch = co_await this->container().invoke_on(
+              controller_stm_shard,
+              &cluster_epoch_generator::get_current_epoch);
+            vassert(
+              new_epoch >= _cached_epoch,
+              "epochs must monotonically increase, but new epoch {} is less "
+              "than the cached epoch of {}",
+              new_epoch,
+              _cached_epoch);
+            _cached_epoch = new_epoch;
+            _cached_epoch_time = ss::lowres_clock::now();
+        }
+    }
+    co_return _cached_epoch;
```

Comment:
so this is to avoid poking core 0 unnecessarily frequently?

Reply:
yes or making cross core calls